diff --git a/attacher.c b/attacher.c
index 18ba43c..097a14e 100644
--- a/attacher.c
+++ b/attacher.c
@@ -891,7 +891,11 @@ screen_builtin_lck()
     {
 #endif
   using_pam = 0;
+#ifndef __MVS__
   pass = ppp->pw_passwd;
+#else
+  pass = "";
+#endif
   if (pass == 0 || *pass == 0)
     {
       if ((pass = getpass("Key:   ")))
@@ -927,14 +931,20 @@ screen_builtin_lck()
           sleep(2);
           return;
 	}
+#ifndef __MVS__
       pass = ppp->pw_passwd = SaveStr(pass);
+#endif
     }
 #ifdef USE_PAM
     }
 #endif
 
   debug("screen_builtin_lck looking in gcos field\n");
+#ifndef __MVS__
   strncpy(fullname, ppp->pw_gecos, sizeof(fullname) - 9);
+#else
+  strncpy(fullname, "N/A",  sizeof(fullname) - 9);
+#endif
   fullname[sizeof(fullname) - 9] = 0;
 
   if ((cp1 = index(fullname, ',')) != NULL)
diff --git a/osdef.h.in b/osdef.h.in
index 6ddbd66..074c0c1 100644
--- a/osdef.h.in
+++ b/osdef.h.in
@@ -153,7 +153,7 @@ extern int   socket __P((int, int, int));
 extern int   connect __P((int, struct sockaddr *, int));
 extern int   bind __P((int, struct sockaddr *, int));
 extern int   listen __P((int, int));
-#ifdef _AIX
+#if defined(__MVS__)
 extern int   accept __P((int, struct sockaddr *, socklen_t *));
 #else
 extern int   accept __P((int, struct sockaddr *, int *));
diff --git a/process.c b/process.c
index 004b22a..c632a70 100644
--- a/process.c
+++ b/process.c
@@ -46,7 +46,7 @@
 
 /* for solaris 2.1, Unixware (SVR4.2) and possibly others: */
 #ifdef HAVE_STROPTS_H
-# include <sys/stropts.h>
+# include <stropts.h>
 #endif
 
 #include "screen.h"
@@ -4055,7 +4055,9 @@ int key;
 	    break;
 	  ApplyAttrColor(i, &mchar_so);
 	  WindowChanged((struct win *)0, 0);
+#ifndef __MVS__
 	  debug2("--> %x %x\n", mchar_so.attr, mchar_so.color);
+#endif
 	}
       if (msgok)
 #ifdef COLOR
